### PR TITLE
[Ruby 2.5] Kernel#warn(..., uplevel:n) 

### PR DESCRIPTION
**Pull request by andrykonchin**
Issue https://github.com/ruby/spec/issues/565
Feature https://bugs.ruby-lang.org/issues/12882

### CODE REVIEW

File: core/kernel/warn_spec.rb [hunk 1]
```diff
@@ -76,4 +76,51 @@
...
+
+    it "raises ArgumentError if passed negative value" do
+      -> { warn "", uplevel: -2 }.should raise_error(ArgumentError)
+      -> { warn "", uplevel: -100 }.should raise_error(ArgumentError)
+    end
+
+    it "processes value -1 like 0" do
+      w = KernelSpecs::WarnInNestedCall.new
+      -> { w.f4("foo", -1) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: foo|)
+    end
```
Comment (andrykonchin) [inline core/kernel/warn_spec.rb]: @eregon I would say it's a bug

File: core/kernel/warn_spec.rb [hunk 2]
```diff
@@ -76,4 +76,53 @@
...
+
+      it "raises ArgumentError if passed negative value" do
+        -> { warn "", uplevel: -2 }.should raise_error(ArgumentError)
+        -> { warn "", uplevel: -100 }.should raise_error(ArgumentError)
+      end
+
+      it "processes value -1 like 0" do
+        w = KernelSpecs::WarnInNestedCall.new
+        -> { w.f4("foo", -1) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: foo|)
+      end
```
Comment (andrykonchin) [inline core/kernel/warn_spec.rb]: @eregon I would say it's a bug

File: core/kernel/warn_spec.rb [hunk 3]
```diff
@@ -76,4 +76,57 @@
...
+
+        -> { w.f4("foo", 0) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 1) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f1_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 2) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f2_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 3) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f3_call_lineno}: warning: foo|)
+      end
+
+      it "does not prepend anything if line number is too big" do
+        w = KernelSpecs::WarnInNestedCall.new
+        -> { w.f4("foo", 100) }.should output(nil, "foo\n")
```
Comment (eregon) [inline core/kernel/warn_spec.rb]: Interestingly this actually crashes with the latest CRuby trunk, good find!

File: core/kernel/warn_spec.rb (hunk 2 above)
Comment (eregon) [reply core/kernel/warn_spec.rb]: Sorry, I missed this message until now (I was quite busy at RubyKaigi).
What do you think should be the behavior? ArgumentError like above?

Comment (eregon) [reply core/kernel/warn_spec.rb]: This was fixed in MRI trunk, it raises an ArgumentError now:
```
$ ruby -e 'warn "", uplevel: -1'
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `warn': negative level (-1) (ArgumentError)
```

Comment (eregon) [reply core/kernel/warn_spec.rb]: I filed https://bugs.ruby-lang.org/issues/14846 to track the backport to 2.5.x.

**Pull request merged**